src/Core/Challenge/Dns/DnspodSolver.php

```diff
@@ -0,0 +1,185 @@
+use AcmePhp\Core\Challenge\MultipleChallengesSolverInterface;
+use AcmePhp\Core\Protocol\AuthorizationChallenge;
+use GuzzleHttp\Client;
+use function GuzzleHttp\json_decode;
+use GuzzleHttp\ClientInterface;
+use Psr\Log\LoggerAwareTrait;
+use Psr\Log\NullLogger;
+use QcloudApi;
+use Webmozart\Assert\Assert;
+
+/**
+ * ACME DNS solver with automate configuration of a DnsPod.cn (TencentCloud NS)
+ *
+ * @author Xiaohui Lam <xiaohui.lam@e.hexdata.cn>
+ */
+class DnspodSolver implements MultipleChallengesSolverInterface, ConfigurableServiceInterface
+{
+    use LoggerAwareTrait;
+    /**
+     * @var DnsDataExtractor
+     */
+    private $extractor;
+
+    /**
+     * @var ClientInterface
+     */
+    private $client;
+
+    /**
+     * @var array
+     */
+    private $cacheZones;
+
+    /**
+     * @var string
+     */
+    private $secretId;
+
+    /**
+     * @var string
+     */
+    private $secretKey;
+
+    /**
+     * @param DnsDataExtractor $extractor
+     * @param ClientInterface  $client
+     */
+    public function __construct(
+        DnsDataExtractor $extractor = null,
+        ClientInterface $client = null
+    ) {
+        $this->extractor = null === $extractor ? new DnsDataExtractor() : $extractor;
+        $this->client = null === $client ? new Client() : $client;
+        $this->logger = new NullLogger();
+    }
+
+    /**
+     * Configure the service with a set of configuration.
+     *
+     * @param array $config
+     */
+    public function configure(array $config)
+    {
+        $this->secretId = $config['secret_id'];
+        $this->secretKey = $config['secret_key'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(AuthorizationChallenge $authorizationChallenge)
+    {
+        return 'dns-01' === $authorizationChallenge->getType();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function solve(AuthorizationChallenge $authorizationChallenge)
+    {
+        return $this->solveAll([$authorizationChallenge]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function solveAll(array $authorizationChallenges)
+    {
+        Assert::allIsInstanceOf($authorizationChallenges, AuthorizationChallenge::class);
+
+        $config = array(
+            'SecretId' => $this->secretId,
+            'SecretKey' => $this->secretKey,
+            'RequestMethod' => 'GET',
+            'DefaultRegion' => 'gz'
+        );
+        /**
+         * @var \QcloudApi_Module_Cns
+         */
+        $cns = QcloudApi::load(QcloudApi::MODULE_CNS, $config);
+
+        foreach ($authorizationChallenges as $authorizationChallenge) {
+            $topLevelDomain = $this->getTopLevelDomain($authorizationChallenge->getDomain());
+            $recordName = $this->extractor->getRecordName($authorizationChallenge);
+            $recordValue = $this->extractor->getRecordValue($authorizationChallenge);
+
+            $subDomain = \str_replace('.'.$topLevelDomain.'.', '', $recordName);
```

Comment:
what if the sub-domain contains the tld? like `foo.io.bar.foo.io`? You should only remove the end of the `$recordName`

Reply:
Yes, seems i missed this situation. thanks.. please wait my re-contribute and re-commit.

Reply:
> what if the sub-domain contains the tld? like `foo.io.bar.foo.io`? You should only remove the end of the `$recordName`

I've done my new code & already simply test